### PR TITLE
Implement Blob URL's DOM interfaces

### PR DESCRIPTION
**Pull request by izgzhen**
r? @Manishearth

Implement the two functions in `URL` to create/revoke Blob URLs, and related code to approximate our proposed design to make things work together.

---

- [x] `./mach build -d` does not report any errors
- [x] `./mach test-tidy` does not report any errors
- [x] These changes fix part of #10539, related to #11131

- [x] There are tests for these changes OR

---

This change is [Reviewable](https://reviewable.io/reviews/servo/servo/11716)

### CODE REVIEW

**Comment by izgzhen** [comment]
I am still not very sure about how to deal with this: the net thread need the `Origin`, where might it come from? The script thread?

There are also other holes in the code, I might fix them later in this PR or following ones.

**Comment by Manishearth** [comment]
For now just send it from the script thread and add a comment saying that net should actually be determining it from the pipeline id. Fetch/http_loader does this too right now. Its wrong, but I wouldn't bother fixing it for the purpose of this pr. If you want to we can try and fix it once everything else is working, otherwise it is on my list-of-things-I'll-eventually-do :smile:

File: components/script/dom/url.rs [hunk 1]
```diff
@@ -105,6 +112,71 @@ impl URL {
...
+        // XXX: Should we send this?
+
+        let slice = blob.get_slice_or_empty();
+        let bytes = slice.get_bytes();
+
+        let entry = BlobURLStoreEntry {
+            type_string: blob.Type().to_string(),
+            filename: None, // XXX: the filename is currently only in File object now
+            size: blob.Size(),
+            bytes: bytes.to_vec(),
```
Comment (Manishearth) [inline components/script/dom/url.rs]: Shoudlnt we be checking for file blobs? And slice blobs, but you  haven't implemented that yet.

Comment (izgzhen) [inline components/script/dom/url.rs]: Yes, you are right. I haven't yet make `Blob` as powerful as it is expected to. But this field is necessary anyway.

**Comment by izgzhen** [comment]
> For now just send it from the script thread and add a comment saying that net should actually be determining it from the pipeline id. Fetch/http_loader does this too right now. Its wrong, but I wouldn't bother fixing it for the purpose of this pr. If you want to we can try and fix it once everything else is working, otherwise it is on my list-of-things-I'll-eventually-do 😄

Cool. Do we have an issue or something to track this?

**Comment by Manishearth** [comment]
Good idea. https://github.com/servo/servo/issues/11722

**Comment by Manishearth** [comment]
Okay, for now send the origin from script (and leave a comment saying we shouldn't trust it linking to that issue). If you want, we could land this first and focus on file blob uris later, or fix it in this PR itself, either way works for me.

**Comment by izgzhen** [comment]
Okay, I still have several gaps to fill. I am okay with the idea of leave a comment if the `Origin` proposal needs some time. Anyway I will keep tracking it.

File: components/script/dom/url.rs [hunk 2]
```diff
@@ -105,6 +112,90 @@ impl URL {
     pub fn DomainToUnicode(_: GlobalRef, origin: USVString) -> USVString {
         USVString(domain_to_unicode(&origin.0))
     }
+
+    // https://w3c.github.io/FileAPI/#dfn-createObjectURL
+    pub fn CreateObjectURL(global: GlobalRef, blob: &Blob) -> DOMString {
+        /// XXX: Second field is an unicode-serialized Origin, it is a temporary workaround
+        ///      and should not be trusted. See issue https://github.com/servo/servo/issues/11722
```
Comment (izgzhen) [inline components/script/dom/url.rs]: Noted both here and in `BlobURLStoreMsg` definition.

File: components/script/dom/url.rs [hunk 3]
```diff
@@ -105,6 +112,90 @@ impl URL {
...
+                }
+                None => {}
+            },
+            Err(_) => {}
+        }
+    }
+
+    // https://w3c.github.io/FileAPI/#unicodeSerializationOfBlobURL
+    fn unicode_serialization_blob_url(origin: &str, id: &str) -> String {
+        // Step 1, 2
```
Comment (izgzhen) [inline components/script/dom/url.rs]: filled now. I feel like there is no more gaps to fill in this PR topic.

**Comment by Manishearth** [comment]
Do any new wpt tests pass? Do you know how to update expectations?

@bors-servo try

**Comment by izgzhen** [comment]
> Do any new wpt tests pass? Do you know how to update expectations?
Yes, sorry about that I forgot to check this out. I have pushed the new code.

File: components/script/dom/url.rs [hunk 4]
```diff
@@ -105,6 +112,90 @@ impl URL {
...
+    // https://w3c.github.io/FileAPI/#dfn-revokeObjectURL
+    pub fn RevokeObjectURL(global: GlobalRef, url: DOMString) {
+        /*
+            If the url refers to a Blob that has a readability state of CLOSED OR
+            if the value provided for the url argument is not a Blob URL, OR
+            if the value provided for the url argument does not have an entry in the Blob URL Store,
+
+            this method call does nothing. User agents may display a message on the error console.
+
+            XXX: First condition can't be enforced given current design
```
Comment (Manishearth) [inline components/script/dom/url.rs]: Why not? 

Anyway, this step seems to already be  run on the fm thread, sort of.

**Comment by Manishearth** [comment]
looks like the manifest_changed thing failed.

Apply this diff, and r=manishearth

```patch
diff --git a/tests/wpt/metadata/MANIFEST.json b/tests/wpt/metadata/MANIFEST.json
index f7861cb..9689adf 100644
--- a/tests/wpt/metadata/MANIFEST.json
+++ b/tests/wpt/metadata/MANIFEST.json
@@ -36026,6 +36026,12 @@
     "deleted_reftests": {},
     "items": {
       "testharness": {
+        "FileAPI/url/url_xmlhttprequest.html": [
+          {
+            "path": "FileAPI/url/url_xmlhttprequest.html",
+            "url": "/FileAPI/url/url_xmlhttprequest.html"
+          }
+        ],
         "cssom-view/scrolling-no-browsing-context.html": [
           {
             "path": "cssom-view/scrolling-no-browsing-context.html",

```

@bors-servo r- delegate+

**Comment by izgzhen** [comment]
Ahh my bad ... I thought both the ini and the metadata needs to be updated

**Comment by izgzhen** [comment]
For the first failed webgl wpt, on my local machine, I got different reasons each time ... 

```
  │ FAIL [expected PASS] WebGL test #0: at (0, 0) expected: 0,0,0,255 was 37,10,160,255
```
```
  │ FAIL [expected PASS] WebGL test #0: at (0, 0) expected: 0,0,0,255 was 14,8,80,255
```

For the second one, I have updated the test ini, though not very clear about why that will happen in worker

**Comment by Manishearth** [comment]
Don't update the ini for the webgl tests

**Comment by izgzhen** [comment]
@Manishearth My fault ... I should rebase it

**Comment by jdm** [comment]
Heh, now that we use the Fetch code in XHR the fetch-related blob code panics when running related tests. We should fix that.

File: components/script/dom/url.rs (hunk 4 above)
Comment (Manishearth) [inline components/script/dom/url.rs]: Change this comment to mention that the first step is unnecessary, since closed blobs do not exist in the store.

**Comment by Manishearth** [comment]
r=me with that comment fixed

@bors-servo delegate+

**Comment by izgzhen** [comment]
Any idea on why it crashed? I can't reproduce this on my mac :(

**Comment by Manishearth** [comment]
cc @jdm

**Comment by Manishearth** [comment]
This probably means that your blob DOM code panicked in a worker. But I can't repro either.

(Initially I thought it was a JIT bug but I misread the backtrace)

**Comment by Manishearth** [comment]
Is there anything in `createObjectURL` that could crash? It seems like `ipc::channel()` is the only one, but that really shouldn't crash...

**Comment by jdm** [comment]
Worker::PostMessage panicked because the destination channel didn't exist. That happened because the blob: url for the worker source couldn't be loaded, so run_worker_scope returned and the thread exited.

**Comment by jdm** [comment]
@bors-servo: retry
* #11693

**Comment by Manishearth** [comment]
> That happened because the blob: url for the worker source couldn't be loaded

oh, it tried to run a worker with a blob url. I didn't notice that 😄

**Comment by izgzhen** [comment]
@jdm Before I retry, do I have to consider the "Worker::PostMessage panicked" problem?

**Comment by jdm** [comment]
Yes, I would prefer that.

**Comment by izgzhen** [comment]
@jdm I checked the code a bit and found that the worker constructor will load resource (`load_whole_resource`) from the Blob URL, which will finally go to [here](https://github.com/servo/servo/blob/master/components/net/resource_thread.rs#L477).

However, since I have not yet routed the `blob`-kind request to file manager thread, it is supposed to return `NetworkError` or something (which might contribute to the TIMEOUT of worker? not sure), rather than crash out of no clear reason ...

Besides, I can't reproduce this locally, and also I am not ready to route the blob URL request yet (still needs more integration), I personally prefer to check if this PR will break again on homu and maybe submit a new issue for it (if it is unreproducible).

**Comment by jdm** [comment]
@izgzhen It should be enough to use `let _ = ...` instead of the `.unwrap()` in `Worker::PostMessage`, since step 9 of https://html.spec.whatwg.org/multipage/comms.html#dom-messageport-postmessage indicates that a nonexistent communication channel should result in a silent error.

File: components/script/dom/worker.rs [hunk 5]
```diff
@@ -147,7 +147,10 @@ impl WorkerMethods for Worker {
     fn PostMessage(&self, cx: *mut JSContext, message: HandleValue) -> ErrorResult {
         let data = try!(StructuredCloneData::write(cx, message));
         let address = Trusted::new(self);
-        self.sender.send((address, WorkerScriptMsg::DOMMessage(data))).unwrap();
+
+        // NOTE: step 9 of https://html.spec.whatwg.org/multipage/comms.html#dom-messageport-postmessage
+        // indicates that a nonexistent communication channel should result in a silent error.
+        let _ = self.sender.send((address, WorkerScriptMsg::DOMMessage(data)));
```
Comment (izgzhen) [inline components/script/dom/worker.rs]: @jdm changed here.

**Pull request merged**